pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+
+Finally, we go back to the root schema resource, enter the second branch of the
+`anyOf` logic applicator, and follow the remote reference at `/anyOf/1/$ref`
+(highlighted in red) into `https://example.com/epoch`. This external schema is
+by definition a separate schema resource. Therefore, it becomes the new lexical
+scope. This time, both the root schema resource and its nested schema resource
+are out of scope:
+
+![Lexical scope after following a reference accross resources (3)](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-ref-remote-3.webp)
+
+Dynamic Scope
+-------------
+
+To recap, the lexical scope of a schema consists of its enclosing schema
+resource. In comparison, the dynamic scope of a schema consists of the stack of
+schema resources evaluated so far. Coming back to our analogy of a schema as a
+graph, the dynamic scope corresponds to the ordered sequence of nodes that were
+visited by the evaluation process.
+
+Consider the following sequence of examples. In the top left, a root schema
+resource named `https://example.com/person` that declares two nested schema
+resources: `https://example.com/name` (at `/properties/name`) and
+`https://example.com/age` (at `/properties/age`). In the bottom left, an
+example instance that successfully validates against the schema. Note that the
+instance does not declare the `age` optional property. In the right, a directed
+graph representation of the relationship between these schema resources.
+Similar to how we did before, we gray out the parts of the schema and of the
+directed graph that are not part of the dynamic scope.
+
+The evaluation process starts with the top level schema. The dynamic scope at
+that point is the root schema resource, and the nested schema resources are out
+of scope. So far the lexical and dynamic scope align:
+
+![The dynamic scope of a JSON Schema (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-1.webp)
+
+Because the instance defines a `name` property, we enter the `properties`
+applicator into the subschema at `/properties/name`. This subschema introduces
+a new schema resource. Therefore, the dynamic scope now consists of *both* the
+root schema resource and the nested schema resource called
+`https://example.com/name`, in order:
+
+![The dynamic scope of a JSON Schema (2)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-2.webp)
+
+In comparison to the lexical scope, the dynamic scope of a schema cannot always
+be statically determined, as the evaluation path often depends on the instance.
+For example, for schemas that make use of logic applicator keywords such as
+`if` or `oneOf`, the ordered sequence of schema resources in scope may vary
+depending on the characteristics of the instance.
+
+### Following References
+
+So far, we learnt that for the lexical scope, following a reference consists in
```

Comment:
Are we going with American English for posts?  If so, then "learnt" should be "learned".

Also a couple other suggestions.

```suggestion
So far, we've learned that for the lexical scope, following a reference consists of
```

Reply:
I guess my English is completely mixed up across regions at this point 😅